### ISSUE TITLE
False warning regarding unsaved files

### ISSUE THREAD

**Issue by GregSal** [microsoft/vscode#236180]
### Applies To

- [x] Notebooks (.ipynb files)
- [ ] Interactive Window and\/or Cell Scripts (.py files with \#%% markers)

### What happened?

VS Code indicating that a file is unsaved when all files are actually saved.
It appear to only happen when Jupyter Notebooks are open.
![Image](https://github.com/user-attachments/assets/61340c1e-68e7-42c1-bcfb-1f0b6cc5004d)

### VS Code Version

1.96.0

### Jupyter Extension Version

2024.11.0

### Jupyter logs

```shell
N/A
```

### Coding Language and Runtime Version

Python v3.11.9  Jupyterlab v4.3.2 cupyter_core v5.72

### Language Extension Version (if applicable)

Python Extension

### Anaconda Version (if applicable)

24.1.2

### Running Jupyter locally or remotely?

Local

**Comment by GregSal** [comment]
Further tests have not replicated the issue on any other PC, but the on the PC where I noticed it, VS code showed an unsaved file even when no files were open in the editor.

I was finally able to clear the problem by closing the workspace.  When I closed the workspace it prompted me to save a file that was not open.  When I selected "Save" it told me that the file in the editor was older than the file in the local folder and asked me if I wanted to overwrite it.  I selected "Overwrite" and the "Unsaved" file warning disappeared.  After closing and re-opening the workspace, Source Control did not show any changes in the "Overwritten" file. and the "Unsaved" file warning did not reappear.

It is worth noting that I had closed and reopened VS code multiple times prior to finding this solution with the "Unsaved" file warning showing up every time I opened this particular workspace, but not when I opened other workspaces.  It just happened to be the case that this was the workspace that used Jupyter notebooks.

I have not been able to reproduce the situation since then, so the problem was likely due to an unclean shutdown of VS code.
If I ever manage to reproduce the problem I will let you know, and attach a copy of the workspace file.  Are there any other files that might also be useful in identifying the problem?

Thanks
Greg Salomons

**Comment by DonJayamanne** [comment]
> I have encountered the issue again on two different occasions on two different PCs with different profiles.

Can you try to disable all extensions and see if this issue repros. 

> VSCode due to a bug in my code that hung the IDE. In the second case it seemed to appear randomly.

Hanging VS Code could have caused a few other issues, hence interested in the latter case.

> When I try to close the workspace I get a prompt to save the .ipynb file followed by a warning that the file is older than the saved version. If I

Are you editing this file using some other application like Jupyter Lab/notebook or another application. 
Are you using remote SSH/docker containers or the like?

**Comment by GregSal** [comment]
> > I have encountered the issue again on two different occasions on two different PCs with different profiles.
> 
> Can you try to disable all extensions and see if this issue repros.

Once a phantom unsaved file appears disabling all extensions does not resolve the issue.  Clearing the phantom unsaved file  required deleting VS Code workspace cache files in APPDATA (windows)

I have not yet reproduced the problem with all extensions disabled, but the amount of testing is limited because the problem is intermittent and there is very little work I can do with all extensions disabled.

> > VSCode due to a bug in my code that hung the IDE. In the second case it seemed to appear randomly.
> 
> Hanging VS Code could have caused a few other issues, hence interested in the latter case.
> 
> > When I try to close the workspace I get a prompt to save the .ipynb file followed by a warning that the file is older than the saved version. If I
> 
> Are you editing this file using some other application like Jupyter Lab/notebook or another application. Are you using remote SSH/docker containers or the like?

All edits were done within VS Code.  Containers were not in use for any the occurrences. In some cases the repository was on a network drive, in other cases on a local hard drive.  In all cases the phantom unsaved file was reported as one of the recently edited jupyter notebooks in the repository.   When all open files were closed VS Code would continue to indicate an unsaved file, but would not specify the file.  In all cases VS Code was being run in a Windows 10 environment.

I will re-open the issue if I encounter any new information that may help.  Reply to this if you would like me to try something specific.

**Comment by ChadBailey** [comment]
Hello, I also experience this frequently and have for a very long time now. It seems to be triggered by resuming an in-progress session particularly over remote ssh.

That said, I think I know a feature that would both fix this AND be a killer feature in VS Code for all users.

I think the thing on all of our minds when we see this is "What changed? Did I begin work here and forget about it or accidentally paste something important/sensitive here?" A way we can determine this is manually open the file on disk using some other tool (haven't found a way to do this natively in vs code), then paste both on-disk and in memory copies into a diff engine such as mergely.com

I think a super elegant solution would be to add a "Show diff" option to these buttons:

![Image](https://github.com/user-attachments/assets/243d1527-02b1-4532-abba-336f5be590eb)

The way I imagine this working is a merge conflict resolution style window opening showing the diff between the file as on disk and the file as in memory. when you hit complete, it will write whatever is in the bottom window to disk the same way a merge conflict resolution window does.

**Comment by amunger** [comment]
@ChadBailey I believe the timeline view in the explorer could be used for that, comparing with the last time the file was saved:

![Image](https://github.com/user-attachments/assets/ab0ed2f3-657e-4f4d-b9d2-36e5c03a5349)

Though it sounds like the original this issue is about a file that somehow gets marked dirty even though there are no changes.

**Comment by ChadBailey** [comment]
You're right, I should have created a new issue and made a reference back to this issue instead.

My +1 still applies here though, I do experience this frequently. Strongest indicator is when you hit Ctrl+z and you have no undo history yet the editor still thinks you have unsaved changes.

**Comment by GregSal** [comment]
It is possible that it is the same issue.  I haven't been getting the problem recently, but I also haven't been using remote ssh recently and I was using it frequently at the time I first reported this..
I didn't make the connection with remote ssh, because the files with the unsaved warning were local not remote.

The only time I have encountered it recently was when a miss-click started to launch an ipython session, which I closed before it finished opening.

**Comment by GregSal** [comment]
I have not encountered this issue in the past few months.
